tower-http/src/timeout/body.rs

```diff
@@ -0,0 +1,347 @@
+//! Middleware that applies a timeout to bodies.
+//!
+//! Bodies must produce data at most within the specified timeout.
+//! If they are inactive, an error will be generated.
+//!
+//! # Differences from [`tower_http::timeout::service::Timeout`]
+//!
+//! [`tower_http::timeout::service::Timeout`] applies a timeout on the full request.
+//! That timeout is not reset when bytes are handled, whether the request is active or not.
+//!
+//! This middleware will return a [`TimeoutError`].
+//!
+//! # Example
+//!
+//! ```
+//! use http::{Request, Response};
+//! use hyper::Body;
+//! use std::time::Duration;
+//! use tower::ServiceBuilder;
+//! use tower_http::timeout::body::RequestTimeoutBodyLayer;
+//!
+//! async fn handle(_: Request<Body>) -> Result<Response<Body>, Infallible> {
+//!     // ...
+//!     # Ok(Response::new(Body::empty()))
+//! }
+//!
+//! # #[tokio::main]
+//! # async fn main() -> Result<(), Box<dyn std::error::Error>> {
+//! let svc = ServiceBuilder::new()
+//!     // Timeout bodies after 30 seconds of inactivity
+//!     .layer(RequestTimeoutBodyLayer::new(Duration::from_secs(30)))
+//!     .service_fn(handle);
+//! # Ok(())
+//! # }
+//! ```
+
+use std::{time::Duration, task::{Context, Poll}, pin::Pin};
+use http_body::Body;
+use futures_core::Future;
+use pin_project_lite::pin_project;
+use tokio::time::{Sleep, sleep};
+use http::{Request, Response};
+use tower_layer::Layer;
+use tower_service::Service;
+
+pin_project! {
+    /// Wrapper around a [`http_body::Body`] to time out if data is not ready within the specified duration.
+    pub struct TimeoutBody<B> {
```

Comment:
Taking a big step back, this _could_ live in `http_body` crate if there was a standard interface for sleep futures.